File: src/main/java/org/cip4/xjdf/json/openapi/TypeMap.java
package org.cip4.xjdf.json.openapi;

import lombok.Getter;
import org.cip4.xjdf.json.openapi.model.Schema;
import org.cip4.xjdf.json.openapi.model.Schemas;

import java.util.HashMap;
import java.util.Map;

@Getter
public class TypeMap {

    private final Map<String, String> prefixes = new HashMap<>();

    public TypeMap(Schemas schemas) {
        schemas.forEach((s, schema) -> register(schemas.getPrefix(), s, schema));
    }

    private void register(String prefix, String rootElement, Schema schema) {
        prefixes.put(rootElement, prefix);
        if (schema.defs() != null) {
            schema.defs().keySet().forEach(key -> prefixes.put(key, prefix));
        }
    }

    public Schema getReference(String elementName) {
        if (!prefixes.containsKey(elementName)) {
            throw new IllegalArgumentException("No reference found for element name: " + elementName);
        }
        return new Schema().ref(prefixes.get(elementName) + elementName);
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/TypeMap.kt
package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.Schema

class TypeMap(schemas: Map<String, Schema>) {

    private val types: MutableMap<String, String> = mutableMapOf()

    init {
        schemas.forEach { register(it.key, it.value) }
    }

    private fun register(rootElement: String, schema: Schema) {
        types[rootElement] = schema.`$id`!!
        val prefix = schema.`$id` + "#/\$defs/"
        schema.`$defs`.keys.forEach { elementName ->
            types[elementName] = prefix.plus(elementName)
        }
    }

    fun getReference(elementName: String): Schema {
        return Schema(
            `$ref` = types[elementName]
        )
    }

}